language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java

```diff
@@ -0,0 +1,109 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.codeaction.providers.changetype;
+
+import io.ballerina.compiler.api.Types;
+import io.ballerina.compiler.api.symbols.TypeDescKind;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.api.symbols.UnionTypeSymbol;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.TemplateExpressionNode;
+import io.ballerina.tools.diagnostics.Diagnostic;
+import org.ballerinalang.annotation.JavaSPIService;
+import org.ballerinalang.langserver.codeaction.CodeActionContextTypeResolver;
+import org.ballerinalang.langserver.codeaction.CodeActionNodeValidator;
+import org.ballerinalang.langserver.codeaction.CodeActionUtil;
+import org.ballerinalang.langserver.common.constants.CommandConstants;
+import org.ballerinalang.langserver.common.utils.PositionUtil;
+import org.ballerinalang.langserver.commons.CodeActionContext;
+import org.ballerinalang.langserver.commons.codeaction.spi.DiagBasedPositionDetails;
+import org.ballerinalang.langserver.commons.codeaction.spi.DiagnosticBasedCodeActionProvider;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.CodeActionKind;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+/***
+ * Code Action for adding subtype for a raw template.
+ * @since 2201.2.1
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class ConvertToXmlOrStringTemplateCodeAction implements DiagnosticBasedCodeActionProvider {
+
+    public static final String NAME = "Change subtype of a raw template";
+    public static final String DIAGNOSTIC_CODE = "BCE3936";
+
+    @Override
+    public boolean validate(Diagnostic diagnostic, DiagBasedPositionDetails positionDetails,
+                            CodeActionContext context) {
+        return DIAGNOSTIC_CODE.equals(diagnostic.diagnosticInfo().code()) &&
+                CodeActionNodeValidator.validate(context.nodeAtRange()) && context.currentSemanticModel().isPresent()
+                && positionDetails.matchedNode().kind() == SyntaxKind.RAW_TEMPLATE_EXPRESSION;
+    }
+
+    @Override
+    public List<CodeAction> getCodeActions(Diagnostic diagnostic,
+                                           DiagBasedPositionDetails positionDetails,
+                                           CodeActionContext context) {
+        TemplateExpressionNode templateExpressionNode = (TemplateExpressionNode) positionDetails.matchedNode();
+        Optional<TypeSymbol> typeSymbol = templateExpressionNode.apply(new CodeActionContextTypeResolver(context));
+
+        Set<String> typeSet = new HashSet<>();
+        if (typeSymbol.isPresent()) {
```

Comment:
Shall we use an early return if typeSymbol is not present?

Reply:
In here the typeSymbol present means we can get the expected type of the raw template. There are cases where we cannot determine the expected type of the raw template. In those case we provide two code actions. Hence we cannot return an empty list even if the typeSymbol is not present.

example:
```
function typeSymbolNotPresent() {
     return `Ballerina`;
}
```

Reply:
@TharushiJay refer #38242 PR for the changes.